packages/keyring-controller/src/KeyringController.ts

```diff
@@ -1228,40 +1717,336 @@ export class KeyringController extends BaseController<
+      await keyring.init();
+    }
+
+    return keyring;
+  }
+
+  /**
+   * Clear Keyrings
+   *
+   * Deallocates all currently managed keyrings and accounts.
+   * Used before initializing a new vault and after locking
+   * MetaMask.
+   */
+  async #clearKeyrings() {
+    // clear keyrings from memory
+    for (const keyring of this.#keyrings) {
+      await this.#destroyKeyring(keyring);
+    }
+    this.#keyrings = [];
+    this.update((state) => {
+      state.keyrings = [];
+    });
+  }
+
+  /**
+   * Restore Keyring Helper
+   *
+   * Attempts to initialize a new keyring from the provided serialized payload.
+   * On success, returns the resulting keyring instance.
+   *
+   * @param serialized - The serialized keyring.
+   * @param serialized.type - Keyring type.
+   * @param serialized.data - Keyring data.
+   * @returns The deserialized keyring or undefined if the keyring type is unsupported.
+   */
+  async #restoreKeyring(
+    serialized: SerializedKeyring,
+  ): Promise<EthKeyring<Json> | undefined> {
+    const { type, data } = serialized;
+
+    let keyring: EthKeyring<Json> | undefined;
+    try {
+      keyring = await this.#newKeyring(type, data);
+    } catch (error) {
+      // Ignore error.
+      console.error(error);
+    }
+
+    if (!keyring) {
+      this.#unsupportedKeyrings.push(serialized);
+      return undefined;
+    }
+
+    // getAccounts also validates the accounts for some keyrings
+    await keyring.getAccounts();
+    this.#keyrings.push(keyring);
+    return keyring;
   }
 
   /**
-   * Handle keyring unlock event.
+   * Destroy Keyring
+   *
+   * Some keyrings support a method called `destroy`, that destroys the
+   * keyring along with removing all its event listeners and, in some cases,
+   * clears the keyring bridge iframe from the DOM.
+   *
+   * @param keyring - The keyring to destroy.
+   */
+  async #destroyKeyring(keyring: EthKeyring<Json>) {
+    await keyring.destroy?.();
+  }
+
+  /**
+   * Remove empty keyrings.
+   *
+   * Loops through the keyrings and removes the ones with empty accounts
+   * (usually after removing the last / only account) from a keyring.
+   */
+  async #removeEmptyKeyrings(): Promise<void> {
+    const validKeyrings: EthKeyring<Json>[] = [];
+
+    // Since getAccounts returns a Promise
+    // We need to wait to hear back form each keyring
+    // in order to decide which ones are now valid (accounts.length > 0)
+
+    await Promise.all(
+      this.#keyrings.map(async (keyring: EthKeyring<Json>) => {
+        const accounts = await keyring.getAccounts();
+        if (accounts.length > 0) {
+          validKeyrings.push(keyring);
+        } else {
+          await this.#destroyKeyring(keyring);
+        }
+      }),
+    );
+    this.#keyrings = validKeyrings;
+  }
+
+  /**
+   * Checks for duplicate keypairs, using the the first account in the given
+   * array. Rejects if a duplicate is found.
+   *
+   * Only supports 'Simple Key Pair'.
```

Comment:
Uhmm, we probably want to avoid account duplication with any type, but I'm wondering if we want to change this behavior in this PR

Reply:
Ah perhaps not, but good eye. I wonder if this was done intentionally for simplicity or was added before the other keyring types were added and hasn't been updated since.